eth/consensus/clique/clique.py

```diff
@@ -0,0 +1,173 @@
+import logging
+import time
+from typing import Sequence, Iterable
+
+from eth.abc import (
+    AtomicDatabaseAPI,
+    BlockHeaderAPI,
+    VirtualMachineAPI,
+)
+from eth.db.chain import ChainDB
+
+from eth_typing import (
+    Address,
+    Hash32,
+)
+from eth_utils import (
+    encode_hex,
+    to_tuple,
+    ValidationError,
+)
+
+from eth.rlp.headers import BlockHeader
+
+from eth.typing import (
+    HeaderParams,
+    VMConfiguration,
+    VMFork,
+)
+from eth.vm.chain_context import ChainContext
+from eth.vm.execution_context import (
+    ExecutionContext,
+)
+
+from .datatypes import (
+    Snapshot,
+)
+from .header_cache import HeaderCache
+from .snapshot_manager import SnapshotManager
+from ._utils import (
+    get_block_signer,
+    is_in_turn,
+    validate_header_integrity,
+)
+
+# Clique Consensus implementation according to
+# https://github.com/ethereum/EIPs/issues/225
+
+
+def create_execution_context(self: BlockHeaderAPI,
+                             prev_hashes: Iterable[Hash32],
+                             chain_context: ChainContext,
+                             import_block: BlockHeaderAPI=None) -> ExecutionContext:
+
+    # In Clique consensus, the tx fee goes to the signer
+    fee_recipient = None
+    if import_block is not None:
+        try:
+            fee_recipient = get_block_signer(import_block.header)
+        except Exception:
+            pass
+
+    return ExecutionContext(
+        coinbase=self.coinbase,
+        timestamp=self.timestamp,
+        block_number=self.block_number,
+        difficulty=self.difficulty,
+        gas_limit=self.gas_limit,
+        prev_hashes=prev_hashes,
+        chain_id=chain_context.chain_id,
+        fee_recipient=fee_recipient
+    )
+
+
+BlockHeader.create_execution_context = create_execution_context  # type: ignore
+
+
+class CliqueConsensus:
+
+    logger = logging.getLogger('eth.consensus.clique.CliqueConsensus')
+
+    def __init__(self, base_db: AtomicDatabaseAPI, epoch_length: int = 30000) -> None:
+        if base_db is None:
+            raise ValueError("Can not instantiate without `base_db`")
+        self._epoch_length = epoch_length
+        self._chain_db = ChainDB(base_db)
+        self._header_cache = HeaderCache(self._chain_db)
+        self._snapshot_manager = SnapshotManager(
+            self._chain_db,
+            self._header_cache,
+            self._epoch_length,
+        )
+
+    @to_tuple
+    def amend_vm_configuration(self, config: VMConfiguration) -> Iterable[VMFork]:
+        for pair in config:
+            block_number, vm = pair
+            vm_class = vm.configure(
+                block_extra_data_max_bytes=65535,
+                validate_seal=lambda header: self.validate_seal(header),
+                configure_header=self.configure_header,
+                _assign_block_rewards=lambda _, __: None,
+            )
```

Comment:
@pipermerriam while you are looking at this, can I drive your attention towards this :point_up: This is basically the heart of how the clique consensus currently fits into the existing architecture. We basically just overwrite a few methods on each VM. This needs to happen early in the process. Notice that these patched methods are operating on an *instance* of `CliqueConsensus` which in turn depends on a database instance.

I'm not sure yet how to replace this with something that doesn't look and feel like monkey patching :see_no_evil:

Reply:
We don't have a solid pattern for doing stuff like this but it would be nice if we could figure one out.

Here's the other places that I know of where we do something similar:

https://github.com/ethereum/py-evm/blob/e8f7039b33177118139bcedcf5570a6bcdecea5e/eth/tools/builder/chain/builders.py#L271-L274

https://github.com/ethereum/py-evm/blob/e8f7039b33177118139bcedcf5570a6bcdecea5e/eth/tools/builder/chain/builders.py#L303-L314

All of them follow roughly the same pattern which suggests that this *could* be made generic, but it would still be a bunch of dynamic class creation which isn't ideal...

Reply:
Maybe it's ok to leave it as is for this PR at least.